### PR TITLE
Feature : create component from template

### PR DESCRIPTION
**Pull request by justinehell**
Close https://github.com/ditrit/leto-modelizer/issues/150

### CODE REVIEW

**Review by Zorin95670 — CHANGES_REQUESTED** [review]

File: tests/unit/composables/Project.spec.js [hunk 1]
```diff
@@ -291,6 +293,18 @@ describe('Test composable: Project', () => {
     });
   });
 
+  describe('Test function: appendProjectFile', () => {
+    it('should succed and return undefined', async () => {
```
[the remark this answers is not in the document]
Comment (justinehell) [reply tests/unit/composables/Project.spec.js:297]: done

File: src/components/card/ComponentDefinitionCard.vue [hunk 2]
```diff
@@ -15,12 +15,33 @@
         avatar
         class="items-center q-pr-none"
       >
+        <template v-if="definition.isTemplate">
+          <q-icon
+            color="primary"
+            size="xl"
+            :title="definition.title"
```
[the remark this answers is not in the document]
Comment (justinehell) [reply src/components/card/ComponentDefinitionCard.vue:22]: done

File: src/components/ModelizerModelView.vue [hunk 3]
```diff
@@ -53,7 +60,27 @@ function updatePlugins() {
   data.plugins.forEach((plugin) => drawComponents(plugin, props.projectName));
 }
 
+/**
+ * Get list of component templates from a remote file. Manage errors.
+ */
+async function getComponentTemplates() {
```
[the remark this answers is not in the document]
Comment (justinehell) [reply src/components/ModelizerModelView.vue:66]: done

File: src/composables/TemplateManager.js [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+}
+
+/**
+ * Filter and return only component templates related to a provided plugin.
+ * @param {Array} templates - Array of metadata structure of templates.
+ * @param {String} pluginName - Name of plugin.
+ * @return {Array<Object>} Filtered component templates.
+ */
+export function setComponentTemplates(templates, pluginName) {
+  return templates
```
[the remark this answers is not in the document]
Comment (justinehell) [reply src/composables/TemplateManager.js:24]: done

File: src/components/drawer/ComponentDefinitionsDrawer.vue [hunk 5]
```diff
@@ -42,6 +42,35 @@
...
+              {{  $t('page.modelizer.drawer.templates.title') }}
+            </q-item-section>
+          </template>
+          <q-scroll-area
+            v-if="templates.length"
+            :style="`height: calc(100vh - ${scrollAreaHeight(definitions)}px);`"
+            class="sunken-area"
+          >
+            <component-definition-grid
+              :definitions="templates"
```
Comment (Zorin95670) [inline src/components/drawer/ComponentDefinitionsDrawer.vue:64]: one line

File: README.md [hunk 6]
```diff
@@ -79,14 +78,40 @@ It will generate the built application in the `dist` folder.
...
+```
+
+To be able to access external resources for your templates, you need to set a reverse proxy named `template-library`. You can see an example below of an `nginx.conf` file:
+
+```bash
+http {
+  server {
+    listen 80;
+    location /template-library {
+      proxy_pass process.env.TEMPLATE_LIBRARY_BASE_URL;
```
Comment (Zorin95670) [inline README.md:107]: proxy pass TEMPLATE_LIBRARY_BASE_URL;

**Review by Zorin95670 — CHANGES_REQUESTED** [review]

File: src/components/drawer/ComponentDefinitionsDrawer.vue (hunk 5 above)
Comment (justinehell) [reply src/components/drawer/ComponentDefinitionsDrawer.vue:64]: done

File: README.md (hunk 6 above)
Comment (justinehell) [reply README.md:107]: done

**Review by k-y0u — CHANGES_REQUESTED** [review]

File: src/composables/Project.js [hunk 7]
```diff
@@ -320,6 +320,30 @@ export async function writeProjectFile(projectId, file) {
   });
 }
 
+/**
+ * Append the given content to a file.
+ * Create the file if not existing.
+ * @param {String} projectId - Id of project.
+ * @param {FileInput} file - File input.
```
[the remark this answers is not in the document]
Comment (justinehell) [reply src/composables/Project.js:327]: done

File: tests/unit/components/ModelizerModelView.spec.js [hunk 8]
```diff
@@ -104,33 +97,14 @@ describe('Test component: ModelizerModelView', () => {
...
-        data: {
-          components: [],
-        },
-        draw: jest.fn(),
-        isParsable: () => true,
-        parse: jest.fn(),
-      };
-
-      await wrapper.vm.drawComponents(plugin);
+  describe('Test functions', () => {
```
[the remark this answers is not in the document]
Comment (justinehell) [reply tests/unit/components/ModelizerModelView.spec.js:91]: done

File: tests/unit/composables/Project.spec.js [hunk 9]
```diff
@@ -291,6 +293,18 @@ describe('Test composable: Project', () => {
...
   });
 
+  describe('Test function: appendProjectFile', () => {
+    it('should succeed and return undefined', async () => {
+      const result = await appendProjectFile('projectId', { path: 'goodPath', content: 'content' });
+      expect(result).toBeUndefined();
+    });
+
+    it('should fail and return error', async () => {
+      const error = await appendProjectFile('projectId', { path: 'error', content: 'content' }).catch((e) => e);
```
[the remark this answers is not in the document]
Comment (justinehell) [reply tests/unit/composables/Project.spec.js:303]: keep it

File: tests/unit/composables/PluginManager.spec.js [hunk 10]
```diff
@@ -172,4 +173,34 @@ describe('Test composable: PluginManager', () => {
...
+      };
+      const result = await PluginManager.getFileInputs(plugin, [{}]);
+
+      expect(Array.isArray(result)).toBeTruthy();
+      expect(result.length).toEqual(1);
+    });
+  });
+
+  describe('Test function: drawComponents', () => {
+    it('should update plugin.components and call draw() function', async () => {
```
[the remark this answers is not in the document]
Comment (justinehell) [reply tests/unit/composables/PluginManager.spec.js:190]: done

File: src/composables/TemplateManager.js [hunk 11]
```diff
@@ -0,0 +1,63 @@
...
+ * Get list of templates from a remote file.
+ * @return {Promise<Array>|Array} Promise with templates on success otherwise an error,
+ * or empty array.
+ */
+export function getRemoteTemplates() {
+  if (process.env.TEMPLATE_LIBRARY_BASE_URL) {
+    return getTemplateFileByPath('index.json')
+      .then((result) => result.data.templates);
+  }
+  return [];
```
[the remark this answers is not in the document]
Comment (justinehell) [reply src/composables/TemplateManager.js:27]: done

File: src/components/ModelizerModelView.vue [hunk 12]
```diff
@@ -45,21 +51,33 @@ const props = defineProps({
...
 
 /**
- * Update plugins array
+ * Update plugins array and related component templates array.
  */
-function updatePlugins() {
+async function updatePluginsAndTemplates() {
   data.plugins = getPlugins();
   data.plugins.forEach((plugin) => drawComponents(plugin, props.projectName));
+  await getTemplatesByType('component', data.plugins[0]?.data?.name)
```
[the remark this answers is not in the document]
Comment (justinehell) [reply src/components/ModelizerModelView.vue:62]: done

File: src/components/card/ComponentDefinitionCard.vue [hunk 13]
```diff
@@ -45,20 +68,51 @@ const props = defineProps({
...
+    return `img:${props.definition.icon}`;
+  }
+  return `img:/plugins/${props.pluginName}/icons/${props.definition.icon}.svg`;
+});
+
 /**
- * On definition click, add a new component to the plugin components
- * and draw them all.
+ * On plugin definition click, add a new component to the plugin components.
+ * On template definition click, get all related remote files and append them to fs.
```
[the remark this answers is not in the document]
Comment (justinehell) [reply src/components/card/ComponentDefinitionCard.vue:84]: done

File: tests/unit/composables/TemplateManager.spec.js [hunk 14]
```diff
@@ -0,0 +1,119 @@
...
+
+jest.mock('src/boot/axios', () => ({
+  templateLibraryApiClient: {
+    get: jest.fn(() => Promise.resolve({ data: 'content' })),
+  },
+}));
+
+describe('Test composable: TemplateManager', () => {
+  describe('Test function: getTemplateFileByPath', () => {
+    it('should return remote file content', async () => {
```
[the remark this answers is not in the document]
Comment (justinehell) [reply tests/unit/composables/TemplateManager.spec.js:13]: done

**Review by k-y0u — APPROVED** [review]

**Review by Zorin95670 — APPROVED** [review]

**Pull request merged**